File: lab3/zad2/go/queue/queue_test.go
package queue

import (
	"crypto/rand"
	"math/big"
	"os"
	"testing"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	"github.com/stretchr/testify/suite"
)

const alphanum = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz"

func generateRandomString(length int) string {
	b := make([]byte, length)
	for i := range b {
		n, _ := rand.Int(rand.Reader, big.NewInt(int64(len(alphanum))))
		b[i] = alphanum[n.Int64()]
	}
	return string(b)
}

type QueueTestSuite struct {
	suite.Suite
}

func TestQueueTestSuite(t *testing.T) {
	suite.Run(t, new(QueueTestSuite))
}

func (s *QueueTestSuite) TestConstructorDestructor() {
	queue1 := NewQueue()
	assert.NotNil(s.T(), queue1)
	assert.NotPanics(s.T(), func() {
		queue1.Print()
	})

	queue2 := NewQueue()
	assert.NotNil(s.T(), queue2)
	assert.NotPanics(s.T(), func() {
		queue2.Print()
	})
}

func (s *QueueTestSuite) TestEnqueueDequeue() {
	queue := NewQueue()
	str1 := generateRandomString(10)
	str2 := generateRandomString(10)
	str3 := generateRandomString(10)

	assert.NotPanics(s.T(), func() {
		queue.Enqueue(str1)
	})
	assert.NotPanics(s.T(), func() {
		queue.Enqueue(str2)
	})
	assert.NotPanics(s.T(), func() {
		queue.Enqueue(str3)
	})
	assert.NotPanics(s.T(), func() {
		queue.Print()
	})

	err := queue.Dequeue()
	assert.NoError(s.T(), err)
	err = queue.Dequeue()
	assert.NoError(s.T(), err)
	err = queue.Dequeue()
	assert.NoError(s.T(), err)
	err = queue.Dequeue()
	assert.Error(s.T(), err)
}

func (s *QueueTestSuite) TestPrintEmpty() {
	queue := NewQueue()
	assert.NotPanics(s.T(), func() {
		queue.Print()
	})
}

func (s *QueueTestSuite) TestFileOperations() {
	filename := "test_queue.txt"
	queue1 := NewQueue()
	queue2 := NewQueue()

	for i := 0; i < 5; i++ {
		str := generateRandomString(10)
		queue1.Enqueue(str)
	}

	outFile, err := os.Create(filename)
	require.NoError(s.T(), err)
	err = queue1.SaveToFile(outFile)
	assert.NoError(s.T(), err)
	outFile.Close()

	inFile, err := os.Open(filename)
	require.NoError(s.T(), err)
	err = queue2.LoadFromFile(inFile)
	assert.NoError(s.T(), err)
	inFile.Close()

	assert.NotPanics(s.T(), func() {
		queue1.Print()
	})
	assert.NotPanics(s.T(), func() {
		queue2.Print()
	})

	os.Remove(filename)
}

func (s *QueueTestSuite) TestBinaryFileOperations() {
	filename := "test_queue_binary.bin"
	queue1 := NewQueue()
	queue2 := NewQueue()

	testStrings := make([]string, 5)
	for i := 0; i < 5; i++ {
		str := generateRandomString(10)
		testStrings[i] = str
		queue1.Enqueue(str)
	}

	outFile, err := os.Create(filename)
	require.NoError(s.T(), err)
	err = queue1.SaveToBinaryFile(outFile)
	assert.NoError(s.T(), err)
	outFile.Close()

	inFile, err := os.Open(filename)
	require.NoError(s.T(), err)
	err = queue2.LoadFromBinaryFile(inFile)
	assert.NoError(s.T(), err)
	inFile.Close()

	assert.NotPanics(s.T(), func() {
		queue1.Print()
	})
	assert.NotPanics(s.T(), func() {
		queue2.Print()
	})

	for range testStrings {
		err := queue2.Dequeue()
		assert.NoError(s.T(), err)
	}

	err = queue2.Dequeue()
	assert.Error(s.T(), err)

	os.Remove(filename)
}

func (s *QueueTestSuite) TestFileOperationsErrors() {
	queue := NewQueue()

	err := queue.SaveToFile(nil)
	assert.Error(s.T(), err)
	assert.Contains(s.T(), err.Error(), "file is nil")

	err = queue.LoadFromFile(nil)
	assert.Error(s.T(), err)
	assert.Contains(s.T(), err.Error(), "file is nil")

	err = queue.SaveToBinaryFile(nil)
	assert.Error(s.T(), err)
	assert.Contains(s.T(), err.Error(), "file is nil")

	err = queue.LoadFromBinaryFile(nil)
	assert.Error(s.T(), err)
	assert.Contains(s.T(), err.Error(), "file is nil")

	filename := "test_nonexistent.txt"
	file, err := os.Open(filename)
	require.Error(s.T(), err)
	if file != nil {
		err = queue.LoadFromFile(file)
		assert.Error(s.T(), err)
		file.Close()
	}

	filename = "test_corrupted.bin"
	corruptedData := []byte{0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00}
	err = os.WriteFile(filename, corruptedData, 0644)
	require.NoError(s.T(), err)

	inFile, err := os.Open(filename)
	require.NoError(s.T(), err)
	err = queue.LoadFromBinaryFile(inFile)
	assert.Error(s.T(), err)
	inFile.Close()

	os.Remove(filename)
}

File: lab3/zad2/cpp/queue/queuetest.cpp
#define BOOST_TEST_MODULE QueueTests
#define BOOST_TEST_DYN_LINK
#include <boost/test/unit_test.hpp>
#include <random>
#include <algorithm>
#include "../../../zad1/cpp/queue.h"

using namespace std;

string generateRandomString(size_t length = 10) {
    static const char alphanum[] =
        "0123456789"
        "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
        "abcdefghijklmnopqrstuvwxyz";
    
    static mt19937 rng(random_device{}());
    uniform_int_distribution<size_t> dist(0, sizeof(alphanum) - 2);
    
    string result;
    result.reserve(length);
    for (size_t i = 0; i < length; ++i) {
        result += alphanum[dist(rng)];
    }
    return result;
}

BOOST_AUTO_TEST_SUITE(QueueTests)

BOOST_AUTO_TEST_CASE(TestConstructorDestructor) {
    Queue queue1;
    BOOST_CHECK_NO_THROW(queue1.print());
    
    Queue* queue2 = new Queue();
    BOOST_CHECK_NO_THROW(queue2->print());
    BOOST_CHECK_NO_THROW(delete queue2);
}

BOOST_AUTO_TEST_CASE(TestEnqueueDequeue) {
    Queue queue;
    string str1 = generateRandomString();
    string str2 = generateRandomString();
    string str3 = generateRandomString();
    
    BOOST_CHECK_NO_THROW(queue.enqueue(str1));
    BOOST_CHECK_NO_THROW(queue.enqueue(str2));
    BOOST_CHECK_NO_THROW(queue.enqueue(str3));
    BOOST_CHECK_NO_THROW(queue.print());
    
    BOOST_CHECK_NO_THROW(queue.dequeue());
    BOOST_CHECK_NO_THROW(queue.dequeue());
    BOOST_CHECK_NO_THROW(queue.dequeue());
    BOOST_CHECK_THROW(queue.dequeue(), runtime_error);
}


BOOST_AUTO_TEST_CASE(TestPrintEmpty) {
    Queue queue;
    BOOST_CHECK_NO_THROW(queue.print());
}

BOOST_AUTO_TEST_CASE(TestFileOperations) {
    const string filename = "test_queue.txt";
    Queue queue1, queue2;
    
    for (int i = 0; i < 5; ++i) {
        string str = generateRandomString();
        queue1.enqueue(str);
    }
    
    ofstream outFile(filename);
    BOOST_REQUIRE(outFile.is_open());
    queue1.saveToFile(outFile);
    outFile.close();
    
    ifstream inFile(filename);
    BOOST_REQUIRE(inFile.is_open());
    queue2.loadFromFile(inFile);
    inFile.close();
    
    BOOST_CHECK_NO_THROW(queue1.print());
    BOOST_CHECK_NO_THROW(queue2.print());
    
    remove(filename.c_str());
}

BOOST_AUTO_TEST_CASE(TestBinaryFileOperations) {
    const string filename = "test_queue_binary.bin";
    Queue queue1, queue2;
    
    vector<string> testStrings;
    for (int i = 0; i < 5; ++i) {
        string str = generateRandomString();
        testStrings.push_back(str);
        queue1.enqueue(str);
    }
    
    ofstream outFile(filename, ios::binary);
    BOOST_REQUIRE(outFile.is_open());
    queue1.saveToBinaryFile(outFile);
    outFile.close();
    
    ifstream inFile(filename, ios::binary);
    BOOST_REQUIRE(inFile.is_open());
    queue2.loadFromBinaryFile(inFile);
    inFile.close();
    
    BOOST_CHECK_NO_THROW(queue1.print());
    BOOST_CHECK_NO_THROW(queue2.print());
    
    for (const auto& str : testStrings) {
        BOOST_CHECK_NO_THROW(queue2.dequeue());
    }
    
    BOOST_CHECK_THROW(queue2.dequeue(), runtime_error);
    
    remove(filename.c_str());
}


BOOST_AUTO_TEST_SUITE_END()